pkg/metrics/metrics.go

```diff
@@ -0,0 +1,53 @@
+package metrics
+
+import (
+	"net"
+	"net/http"
+
+	c "github.com/scality/cosi-driver/pkg/constants"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/klog/v2"
+)
+
+var (
+	RequestsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "cosi_requests_total",
+			Help: "Total number of requests handled by the COSI driver.",
+		},
+		[]string{"method", "status"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(RequestsTotal)
+}
+
+func StartMetricsServer(addr string) (*http.Server, error) {
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return StartMetricsServerWithListener(listener)
+}
+
+func StartMetricsServerWithListener(listener net.Listener) (*http.Server, error) {
+	mux := http.NewServeMux()
+	mux.Handle(c.MetricsPath, promhttp.Handler())
+
+	srv := &http.Server{
+		Handler: mux,
+		Addr:    listener.Addr().String(),
+	}
+
+	go func() {
+		klog.InfoS("Starting Prometheus metrics server", "address", listener.Addr().String())
+		if err := srv.Serve(listener); err != nil && err != http.ErrServerClosed {
```

Comment:
For my own edification, why is `ErrServerClosed` not considered an error?

Reply:
From what I understand, the `http.ErrServerClosed `error is not considered an error in this context because it is a normal part of the lifecycle of an HTTP server in Go.

Bacially, the `http.ErrServerClosed` is returned by the `http.Server.ListenAndServe` method when the server is shut down using the Shutdown or Close methods. This indicates a graceful shutdown, which is expected behavior, not a failure. Treating `http.ErrServerClosed` as a non-error ensures that the shutdown process doesn’t log misleading or unnecessary error messages.

Reply:
Yep makes total sense, thanks!